cargo/cargo_build_script_runner/bin.rs

```diff
@@ -31,118 +31,184 @@ fn main() -> Result<(), String> {
+            .as_bytes(),
+    )
+    .expect(&format!("Unable to write file {:?}", depenvfile));
 
-            let output = BuildScriptOutput::from_command(&mut command).map_err(|exit_code| {
-                format!(
-                    "Build script process failed{}",
-                    if let Some(exit_code) = exit_code {
-                        format!(" with exit code {}", exit_code)
-                    } else {
-                        String::new()
-                    }
-                )
-            })?;
-
-            // The right way to set the dep env var is to use the links attribute from the
-            // Cargo.toml, but cargo_build_script didn't used to have a `links` attribute, so for
-            // backward-compatibility reasons, try to infer it from the name of the crate.
-            // TODO: remove this backward-compatibility fallback in next major version.
-            let crate_links = match crate_links.as_ref() {
-                "" => {
-                    const SYS_CRATE_SUFFIX: &str = "-sys";
-                    if crate_name.ends_with(SYS_CRATE_SUFFIX) {
-                        crate_name
-                            .split_at(crate_name.rfind(SYS_CRATE_SUFFIX).unwrap())
-                            .0
-                    } else {
-                        &crate_name
-                    }
-                },
-                crate_links => crate_links,
-            };
-
-            write(&envfile, BuildScriptOutput::to_env(&output, &exec_root.to_string_lossy()).as_bytes())
-                .expect(&format!("Unable to write file {:?}", envfile));
-            write(&depenvfile, BuildScriptOutput::to_dep_env(&output, crate_links, &exec_root.to_string_lossy()).as_bytes())
-                .expect(&format!("Unable to write file {:?}", depenvfile));
-
-            let CompileAndLinkFlags { compile_flags, link_flags } = BuildScriptOutput::to_flags(&output, &exec_root.to_string_lossy());
-
-            write(&flagfile, compile_flags.as_bytes())
-                .expect(&format!("Unable to write file {:?}", flagfile));
-            write(&linkflags, link_flags.as_bytes())
-                .expect(&format!("Unable to write file {:?}", linkflags));
-            Ok(())
+    let CompileAndLinkFlags {
+        compile_flags,
+        link_flags,
+    } = BuildScriptOutput::to_flags(&output, &exec_root.to_string_lossy());
+
+    write(&flagfile, compile_flags.as_bytes())
+        .expect(&format!("Unable to write file {:?}", flagfile));
+    write(&linkflags, link_flags.as_bytes())
+        .expect(&format!("Unable to write file {:?}", linkflags));
+    Ok(())
+}
+
+/// A representation of expected command line arguments.
+struct Options {
+    progname: String,
+    crate_name: String,
+    crate_links: String,
+    out_dir: String,
+    envfile: String,
+    flagfile: String,
+    linkflags: String,
+    depenvfile: String,
+    remaining_args: Vec<String>,
```

Comment:
Maybe `input_dep_env_paths` to be more descriptive of what it contains? And maybe call `depenvfile` instead `output_dep_env_path` to differentiate?